test/restart/test_uda.py

```diff
@@ -0,0 +1,110 @@
+from cassandra.protocol import InvalidRequest
+
+@pytest.mark.asyncio
+async def test_uda_restart(manager):
+    """Test UDA survives server restart"""
+    servers = await manager.servers()
+    schema = 'pk int PRIMARY KEY, tx text'
+    keyspace = unique_name()
+    manager.cql.execute(f"CREATE KEYSPACE {keyspace} WITH REPLICATION = {{ 'class' : 'SimpleStrategy', 'replication_factor' : 1 }}")
+
+    # Prepare a test table
+    table = keyspace + "." + unique_name()
+    manager.cql.execute(f"CREATE TABLE {table} ({schema})")
+    for x in range(1, 100):
+        s = "a" * x
+        manager.cql.execute(f"INSERT INTO {table} (pk, tx) VALUES ({x}, '{'a' * x}');")
+
+
+    # Create a UDA that sums integers from all rows
+    add = unique_name()
+    manager.cql.execute(f"CREATE FUNCTION {keyspace}.{add} (x int, y int) RETURNS NULL ON NULL INPUT RETURNS int LANGUAGE lua AS 'return x + y'")
+
+    return_input_int = unique_name()
+    manager.cql.execute(f"CREATE FUNCTION {keyspace}.{return_input_int} (acc int) RETURNS NULL ON NULL INPUT RETURNS int LANGUAGE lua AS 'return acc'")
+
+    sum = keyspace + "." + unique_name()
+    manager.cql.execute(f"CREATE AGGREGATE {sum} (int) SFUNC {add} STYPE int FINALFUNC {return_input_int} INITCOND 0")
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum}(pk) FROM {table}")]
+    assert res == [[4950]]
+
+    # Create a variant of the UDA that uses a REDUCEFUNC
+    sum_with_reduce = keyspace + "." + unique_name()
+    manager.cql.execute(f"CREATE AGGREGATE {sum_with_reduce} (int) SFUNC {add} STYPE int REDUCEFUNC {add} FINALFUNC {return_input_int} INITCOND 0")
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum_with_reduce}(pk) FROM {table}")]
+    assert res == [[4950]]
+
+    # Create a UDA that concatenates strings from all rows
+    concat = unique_name()
+    manager.cql.execute(f"CREATE FUNCTION {keyspace}.{concat} (x text, y text) RETURNS NULL ON NULL INPUT RETURNS text LANGUAGE lua AS 'return x..y'")
+
+    return_input_string = unique_name()
+    manager.cql.execute(f"CREATE FUNCTION {keyspace}.{return_input_string} (acc text) RETURNS NULL ON NULL INPUT RETURNS text LANGUAGE lua AS 'return acc'")
+
+    manager.cql.execute(f"CREATE AGGREGATE {sum} (text) SFUNC {concat} STYPE text FINALFUNC {return_input_string} INITCOND ''")
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum}(tx) FROM {table} WHERE pk IN (1,3,5)")]
+    assert res == [['aaaaaaaaa']]
+
+    # Create a variant of the UDA that uses a REDUCEFUNC
+    manager.cql.execute(f"CREATE AGGREGATE {sum_with_reduce} (text) SFUNC {concat} STYPE text REDUCEFUNC {concat} FINALFUNC {return_input_string} INITCOND ''")
+    res = [list(row) for row in manager.cql.execute(f"SELECT {sum_with_reduce}(tx) FROM {table} WHERE pk IN (1,3,5)")]
+    assert res == [['aaaaaaaaa']]
+
+    # Restart server before re-using AGGREGATE
+    await manager.server_restart(servers[0])
+    await manager.driver_connect()
```

Comment:
@kbr- is this what you ended up deciding on? Functions like server_restart() waits for some of the restart to happen, but not enough for the driver to notice, so the caller needs to also wait for "driver_connect()"?
I guess it works, but I think it's risky that test developers will forget this driver_connect() is necessary, and the result will be flaky tests.

Reply:
I couldn't get this to work without performing both `server_restart()` and `driver_connect()`. Luckily we don't have to concern ourselves anymore with this when the test is moved to dtests.